ResearchKit/ActiveTasks/ORKToneAudiometryStepViewController.m

```diff
@@ -0,0 +1,210 @@
+    toneResult.endDate = now;
+    toneResult.samples = [self.samples copy];
+
+    [results addObject:toneResult];
+    sResult.results = [results copy];
+
+    return sResult;
+}
+
+- (void)stepDidFinish {
+    [super stepDidFinish];
+
+    self.expired = YES;
+    [self.toneAudiometryContentView finishStep:self];
+    [self goForward];
+}
+
+- (void)start {
+    [super start];
+
+    [self startCurrentTest];
+}
+
+- (void)suspend {
+    [super suspend];
+    // ...
+}
+
+- (void)resume {
+    [super resume];
+    // ...
+}
+
+- (void)finish {
+    [super finish];
+    // ...
+}
+
+- (IBAction)buttonPressed:(id)button forEvent:(UIEvent *)event {
+    if (self.samples == nil) {
+        _samples = [NSMutableArray array];
+    }
+
+    ORKToneAudiometrySample *sample = [ORKToneAudiometrySample new];
+    NSUInteger frequencyIndex = self.currentTestIndex/2;
+    NSNumber *frequency = self.testingFrequencies[frequencyIndex];
+    sample.frequency = frequency;
+    sample.channel = (self.currentTestIndex%2 == 0) ? ORKAudioChannelLeft : ORKAudioChannelRight;
+    sample.amplitude = @(self.audioGenerator.volumeAmplitude);
+
+    [self.samples addObject:sample];
+
+    [NSRunLoop cancelPreviousPerformRequestsWithTarget:self selector:@selector(testExpired) object:nil];
+    [self.audioGenerator stop];
+
+    [self startNextTestOrFinish];
+}
+
+- (void)testExpired {
+    [self.audioGenerator stop];
+
+    [self startNextTestOrFinish];
+}
+
+- (void)startNextTestOrFinish {
+    self.currentTestIndex ++;
+    if (self.currentTestIndex == (self.testingFrequencies.count * 2)) {
+        [self finish];
+    }
+    else {
+        [self startCurrentTest];
+    }
+}
+
+- (ORKToneAudiometryStep *)toneAudiometryStep {
+    return (ORKToneAudiometryStep *)self.step;
+}
+
+- (void)startCurrentTest {
+    const NSTimeInterval SoundDuration = self.toneAudiometryStep.toneDuration;
+
+    NSUInteger frequencyIndex = self.currentTestIndex/2;
+    if (frequencyIndex >= self.testingFrequencies.count) {
+        return;
+    }
+
+    NSNumber *frequency = self.testingFrequencies[frequencyIndex];
+    ORKAudioChannel channel = (self.currentTestIndex%2 == 0) ? ORKAudioChannelLeft : ORKAudioChannelRight;
+
+    CGFloat progress = 0.001 + (CGFloat)self.currentTestIndex/(self.testingFrequencies.count * 2);
+    [self.toneAudiometryContentView setProgress:progress
+                                        caption:(channel == ORKAudioChannelLeft) ? [NSString stringWithFormat:ORKLocalizedString(@"TONE_LABEL_%@_LEFT", nil), frequency] : [NSString stringWithFormat:ORKLocalizedString(@"TONE_LABEL_%@_RIGHT", nil), frequency]
+                                       animated:YES];
+
+    [self.audioGenerator playSoundAtFrequency:frequency.doubleValue
+                                    onChannel:channel
+                               fadeInDuration:SoundDuration];
+
+    [self performSelector:@selector(testExpired) withObject:nil afterDelay:SoundDuration];
```

Comment:
If using dispatch_after, just maintain an ivar to track whether action is required. On user cancel, reset the ivar. Assuming all work is done on the main queue, this is pretty simple.

Generally with ARC, prefer not to use performSelector: variants if we can avoid it, as it's error prone.